spras/config/container_schema.py

```diff
@@ -0,0 +1,69 @@
+"""
+The separate container schema specification file.
+For information about pydantic, see schema.py.
+
+We move this to a separate file to allow `containers.py` to explicitly take in
+this subsection of the configuration.
+"""
+
+import warnings
+from dataclasses import dataclass
+
+from pydantic import BaseModel, ConfigDict
+
+from spras.config.util import CaseInsensitiveEnum
+
+DEFAULT_CONTAINER_PREFIX = "docker.io/reedcompbio"
+
+class ContainerFramework(CaseInsensitiveEnum):
+    docker = 'docker'
+    # TODO: add apptainer variant once #260 gets merged
+    singularity = 'singularity'
+    dsub = 'dsub'
+
+class ContainerRegistry(BaseModel):
+    base_url: str = "docker.io"
+    "The domain of the registry"
+
+    owner: str = "reedcompbio"
+    "The owner or project of the registry"
+
+    model_config = ConfigDict(extra='forbid', use_attribute_docstrings=True)
+
+class ContainerSettings(BaseModel):
+    framework: ContainerFramework = ContainerFramework.docker
+    unpack_singularity: bool = False
+    registry: ContainerRegistry
+    hash_length: int = 7
+
+    model_config = ConfigDict(extra='forbid')
+
+@dataclass
+class ProcessedContainerSettings:
+    framework: ContainerFramework = ContainerFramework.docker
+    unpack_singularity: bool = False
+    prefix: str = DEFAULT_CONTAINER_PREFIX
+    hash_length: int = 7
```

Comment:
Why is the `hash_length` part of the container settings?

Reply:
This just makes it easier to pass around to `containers.py`, untied to the global configuration. This is also the only 'weird' change, but the file generated by `hash_length` will never appear in `output`.

By default, `hash_length` is set to the top-level hash length.

(This can be added to a doc comment for `hash_length` since this was not clear before)